engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java

```diff
@@ -2424,4 +2435,103 @@ public int hashCode()
     return result;
   }
 
+  /**
+   * Return list of operators in topological sorted order. The Delay operators
+   * are considered as downstream operator.
+   * @return
+   */
+  @VisibleForTesting
+  protected List<OperatorMeta> getSortedOperators()
+  {
+    List<OperatorMeta> sortedOperators = new ArrayList<>(operators.size());
+    Stack<OperatorMeta> pendingNodes = new Stack<>();
+    Set<OperatorMeta> visited = new HashSet<>();
+
+    for (OperatorMeta n : this.getAllOperators()) {
+      pendingNodes.push(n);
+    }
+
+    while (!pendingNodes.isEmpty()) {
+      OperatorMeta n = pendingNodes.pop();
+
+      if (visited.contains(n)) {
+        continue;
+      }
+
+      boolean upstreamVisited = true;
+      for (Map.Entry<InputPortMeta, StreamMeta> entry : n.getInputStreams().entrySet()) {
+        StreamMeta s = entry.getValue();
+        boolean delay = entry.getKey().getValue(LogicalPlan.IS_CONNECTED_TO_DELAY_OPERATOR);
+        // skip delay sources since it's going to be handled as downstream
+        if (!delay && s.getSource() != null && !visited.contains(s.getSource().getOperatorMeta())) {
+          pendingNodes.push(n);
+          pendingNodes.push(s.getSource().getOperatorMeta());
+          upstreamVisited = false;
+          break;
+        }
+      }
+
+      if (upstreamVisited) {
+        visited.add(n);
+        sortedOperators.add(n);
+      }
+    }
+
+    return sortedOperators;
+  }
+
+  /**
+   * Get list of stream from root operator to downstreams. The stream connecting
+   * from delay operators are considered as downstream streams.
+   * @return
+   */
+  @VisibleForTesting
+  protected List<StreamMeta> getSortedStreams()
+  {
+    List<OperatorMeta> sorted = getSortedOperators();
+    return getSortedStreams(sorted);
+  }
+
+  private List<StreamMeta> getSortedStreams(List<OperatorMeta> sorted)
+  {
+    List<StreamMeta> sortedStreams = new ArrayList<>(this.streams.size());
+    for (OperatorMeta meta : sorted) {
+      for (Map.Entry<OutputPortMeta, StreamMeta> entry : meta.getOutputStreams().entrySet()) {
+        StreamMeta s = entry.getValue();
+        sortedStreams.add(s);
+      }
+    }
+    return sortedStreams;
+  }
+
+  /**
+   * Iterate over all registered {@see Visitor} and invoke visit method
+   * for operator and stream. First visitOperator is called for all operators
+   * in the DAG followed by visitStream is called for all the streams in the
+   * DAG.
+   */
+  public void visitDAG()
+  {
+    Collection<Visitor> visitors = getAttributes().get(VISITORS);
+    if (visitors == null) {
+      return;
+    }
+    // First make sure that DAG structure is valid. i.e it does not contain
+    // invalid cycles.
+    validateDAGStructure();
+
+    List<OperatorMeta> sortedOperators = getSortedOperators();
+    List<StreamMeta> sortedStreams = getSortedStreams(sortedOperators);
+    for (Visitor visitor : visitors) {
+      LOG.info("Using visitor {}", visitor);
+      visitor.preVisitDAG(this);
```

Comment:
This lifecycle seems very localized and constrained. I suggest the following lifecycle.

visitor->prePopulateDAG
populateDAG
visitor->postPopulateDAG
visitor->preSetProperties
setProperties and attributes from property file
visitor->postSetProperties
visitor->preValidateDAG - No need for validation of DAG structure or ordering
validate
visitor->postValidateDAG - Let the visitor developer do what they want to do with walking with operators or streams. You can provide convenience methods in the interface to get access to operators and streams in a topographical order)